src/librustc_mir/borrow_check/nll/region_infer/error_reporting/mod.rs

```diff
@@ -120,19 +122,50 @@ impl<'tcx> RegionInferenceContext<'tcx> {
         // we still want to screen for an "interesting" point to
         // highlight (e.g., a call site or something).
         let target_scc = self.constraint_sccs.scc(target_region);
-        let best_choice = (0..path.len()).rev().find(|&i| {
-            let constraint = path[i];
+        let mut range = 0..path.len();
+
+        let should_reverse = match from_region_origin {
```

Comment:
I think we should call this `blame_source`, and we should add a comment like this:

---

As noted above, when reporting an error, there is typically a chain of constraints leading from some "source" region to some "target" region. In most cases, we prefer to "blame" the constraints closer to the target -- but there is one exception. When constraints arise from higher-ranked subtyping, we generally prefer to blame the *source* value, as the "target" in this case tends to be some type annotation that the user gave. Therefore, if we find that the region origin is some instantiation of a higher-ranked region, we start our search from the "source" point rather than the "target", and we also tweak a few other things.

An example might be this bit of Rust code:

```rust
let x: fn(&'static ()) = |_| {};
let y: for<'a> fn(&'a ()) = x;
```

Here, when type-checking, we will instantiate the `'a` with a placeholder region `!a`, which will being expected to outlive `'static`:

```
fn(&'static u32) <: for<'a> fn(&'a u32)
&!a u32 <: &'static u32
!a: 'static
```

Here, we would prefer to blame the source region of the chain (`'a`) and not the target XXX

---

actually, I'm a bit confused about the example, so let me do a build and try to understand what's going on. But regardless I think we should rename this and offer some more detailed explanation of what's going on.

Reply:
OK I dug into what's going on. I'm not 100% convinced this heuristic of reversing source/target will apply super broadly -- it seems like we will ultimately find a more base way to express, but this explanation of what is happening should help us when trying to find that.